fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/AutoContinuingCursor.java

```diff
@@ -0,0 +1,188 @@
+ *         If scan limit properties of the <i>underlying cursor</i> are reached, the generator is asked for a new
+ *         <i>underlying cursor</i> which (1) is in the context of a new transaction and (2) takes the continuation from
+ *         the previous <i>underlying cursor</i>. Then the process is repeated.
+ *     </li>
+ * </ul>
+ *
+ * <p>
+ * {@link AutoContinuingCursor} is responsible for all {@link FDBRecordContext} management, so all reads it does are in
+ * the scope of transactions that it controls.
+ * </p>
+ *
+ * <p>
+ * The {@link AutoContinuingCursor} has no visibility into the {@link com.apple.foundationdb.record.ScanProperties} of
+ * the <i>underlying cursor</i> and, therefore, will not be involved in enforcing any limits that may be individually
+ * applied to the <i>underlying cursor</i>. For example, if the generator returns an <i>underlying cursor</i> that
+ * specified, say, a record scan limit of 10 record, the {@link AutoContinuingCursor} will scan all data until it is
+ * exhausted, at most 10 records at a transaction.
+ * </p>
+ *
+ * @param <T> the type of elements returned by this cursor
+ */
+@API(API.Status.EXPERIMENTAL)
+public class AutoContinuingCursor<T> implements BaseCursor<T> {
+
+    @Nonnull
+    private FDBDatabaseRunner runner;
+    @Nonnull
+    private BiFunction<FDBRecordContext, byte[], RecordCursor<T>> nextCursorGenerator;
+
+    @Nullable
+    private RecordCursor<T> currentCursor;
+    @Nullable
+    private FDBRecordContext currentContext;
+
+    @Nullable
+    private CompletableFuture<Boolean> nextFuture;
+    @Nullable
+    private RecordCursorResult<T> lastResult;
+
+    // for detecting incorrect cursor usage
+    private boolean mayGetContinuation = false;
+
+    /**
+     * Creates a new {@link AutoContinuingCursor}.
+     * @param runner the runner from which it can open new contexts
+     * @param nextCursorGenerator the method which can generate the underlying cursor given a record context and a
+     * continuation
+     */
+    public AutoContinuingCursor(@Nonnull FDBDatabaseRunner runner,
+                                @Nonnull BiFunction<FDBRecordContext, byte[], RecordCursor<T>> nextCursorGenerator) {
+        this.runner = runner;
+        this.nextCursorGenerator = nextCursorGenerator;
+    }
+
+    @Nonnull
+    @Override
+    public CompletableFuture<RecordCursorResult<T>> onNext() {
+        if (currentCursor == null) {
+            openContextAndGenerateCursor(null);
+        }
+        return currentCursor.onNext().thenCompose(recordCursorResult -> {
+            if (recordCursorResult.noNextButHasContinuation()) {
+                openContextAndGenerateCursor(lastResult.getContinuation().toBytes());
+                return currentCursor.onNext().thenApply(finalRecordCursorResult -> {
+                    if (finalRecordCursorResult.noNextButHasContinuation()) {
+                        throw new RecordCoreException("failed to continue the cursor in a new context");
```

Comment:
I'm not 100% sure of the semantics that we want here but this could totally happen in practice and it seems okay to me.

For example, suppose that we have a very selective `FilterCursor` around a `KeyValueCursor` and a relatively low scan limit. In that case, it will be totally normal to have a cursor that we just opened return a `RecordCursorResult` with a continuation and a `NoNextReason` of `SCAN_LIMIT_REACHED`, which seems like it would cause an exception here. Maybe that's what we actually want here, but if it is we should document it carefully.

Alternatively, it seems reasonable to use `AsyncUtils.whileTrue()` to keep generating new cursors (perhaps in the same context?) until we get some kind of result from `currentCursor.onNext()`.

Reply:
My concern was it might be an infinite loop if the generated cursor always return no next given a new cursor. But maybe it is expected.

@scgray What do you think?

Reply:
@nblintao I agree with @nschiefer - we should continue to loop with new contexts as long as the cursor is making progress, whether or not that progress is returning a value.